diff --git a/contrib/systemd/xdg-desktop-portal-hyprland.service.in b/contrib/systemd/xdg-desktop-portal-hyprland.service.in
index a2e3760..b85fb7e 100644
--- a/contrib/systemd/xdg-desktop-portal-hyprland.service.in
+++ b/contrib/systemd/xdg-desktop-portal-hyprland.service.in
@@ -8,5 +8,6 @@ ConditionEnvironment=WAYLAND_DISPLAY
 Type=dbus
 BusName=org.freedesktop.impl.portal.desktop.hyprland
 ExecStart=@LIBEXECDIR@/xdg-desktop-portal-hyprland
-Restart=on-failure
+Restart=always
+RestartSec=1s
 Slice=session.slice
diff --git a/src/core/PortalManager.cpp b/src/core/PortalManager.cpp
index 5dc98d4..87e684c 100644
--- a/src/core/PortalManager.cpp
+++ b/src/core/PortalManager.cpp
@@ -347,6 +347,9 @@ void CPortalManager::startEventLoop() {
 
             int ret = poll(m_sEventLoopInternals.pollFds.data(), m_sEventLoopInternals.pollFds.size(), 5000 /* 5 seconds, reasonable. It's because we might need to terminate */);
             if (ret < 0) {
+                if (errno == EINTR)
+                    continue;
+
                 Debug::log(CRIT, "[core] Polling fds failed with {}", strerror(errno));
                 g_pPortalManager->terminate();
                 break;
@@ -480,18 +483,29 @@ void CPortalManager::startEventLoop() {
 
     Debug::log(ERR, "[core] Terminated");
 
+    if (m_sTimersThread.thread && m_sTimersThread.thread->joinable())
+        m_sTimersThread.thread->join();
+    pollThr.join();
+
     m_sPortals.globalShortcuts.reset();
     m_sPortals.screencopy.reset();
     m_sPortals.screenshot.reset();
-    m_sHelpers.toplevel.reset();
     m_sPortals.inputCapture.reset();
 
+    m_sHelpers.toplevel.reset();
+    m_sHelpers.toplevelMapping.reset();
+
+    m_vOutputs.clear();
+    m_sWaylandConnection.linuxDmabufFeedback.reset();
+    m_sWaylandConnection.linuxDmabuf.reset();
+    m_sWaylandConnection.hyprlandToplevelMgr.reset();
+    m_sWaylandConnection.xdgOutputManager.reset();
+    m_sWaylandConnection.shm.reset();
+    m_sWaylandConnection.registry.reset();
+
     m_pConnection.reset();
     pw_loop_destroy(m_sPipewire.loop);
     wl_display_disconnect(m_sWaylandConnection.display);
-
-    m_sTimersThread.thread.release();
-    pollThr.join(); // wait for poll to exit
 }
 
 sdbus::IConnection* CPortalManager::getConnection() {
diff --git a/src/core/PortalManager.hpp b/src/core/PortalManager.hpp
index eb62c1b..3d5d78e 100644
--- a/src/core/PortalManager.hpp
+++ b/src/core/PortalManager.hpp
@@ -125,8 +125,8 @@ class CPortalManager {
     void  startEventLoop();
     void  setupXDGOutput(SOutput* output);
 
-    bool  m_bTerminate = false;
-    pid_t m_iPID       = 0;
+    std::atomic<bool> m_bTerminate = false;
+    pid_t             m_iPID       = 0;
 
     struct {
         std::condition_variable              loopSignal;
